### ISSUE TITLE
Store ratings and play counts in tags whenever they're modified

### ISSUE THREAD

**Issue by CreamyCookie** [quodlibet/quodlibet#1852]
Given that "Save Ratings and Play Counts" is enabled in the option, whenever the ratings are set, or a song has been played, write rating or play counts to the tags of the song.

Currently one has to use the Force Write plugin, but given the bug reports regarding ratings, most users seem to expect "Save Ratings and Play Counts" to mean that QL stores the ratings automatically in the tags, whenever they change and not just, as is currently the case, when plugins such as ReplayGain or Force Write plugin are used on the songs.

**Comment by nodiscc** [comment]
> Currently one has to use the Force Write plugin, but given the bug reports regarding ratings, most users seem to expect "Save Ratings and Play Counts" to mean that QL stores the ratings automatically in the tags, whenever they change.

I just wanted to confirm this is the behavior that I was expecting. Having to manually "Force Write" is confusing and easy to forget, it has lead me to losing rating data more than once :)

Some users may not want to have their files written to, may I suggest implementing as an option for the `Force write` plugin? I.e. to have ratings saved to file automatically  on change

 * Enable the `Force write` plugin
 * In the plugin preferences, enable `Force write automatically when metadata changes`

**Comment by lazka** [comment]
There was a plan to add a timestamp for when any playback statistic related value was changed (`needs_write = state_timestamp > file_modified_timestamp`). This would allow to write ratings etc at a later point, manually, or do writes of multiple changed files every 10 minutes or so..

**Comment by CreamyCookie** [comment]
> Though some users may not want to have their files forcibly written to. May I suggest implementing as an option for the Force write plugin? I.e. to have ratings saved to file automatically on change

Maybe I'm wrong about this, but I think making this part of the main application would be better, since there's already an option called "Save Ratings and Play Counts" and as, for example, you and me, have said, we expected this to store the ratings and play counts in the tags, not just in the song database. (btw. Does QL really completely stop storing ratings if this is disabled?) Furthermore, plugins currently can't listen to ratings event (#963), so the plugin might have to check this manually over and over again.. I guess play counts could be done with an events plugin, but I'm wondering if this would be really the cleanest solution.

So maybe, to make the distinction between saving them at all, and storing them in the tags more clear, we could add an option beneath "Save Ratings and Play Counts", which states something like "Also store them in the tags" or "Write to files (only supported by mp3, FLAC, and ..)".

**Comment by declension** [comment]
Surely every O/S (or hardware) disk cache should take care of asynchronous writes better than we could.

Either way I'm also keen that this is supported in the core (configurably).

**Comment by CreamyCookie** [comment]
In an EventPlugin, is there a way to check if a rating is already written to the file?

**Comment by XIncognito10** [comment]
any updates on this?

**Comment by zsau** [comment]
I think this would also help my use case. I don't save ratings or playcounts to tags, but still run into the problem that if QL exits abnormally, certain state info gets lost (including last played date as well as view state info such as the current browser selection). Saving more state either immediately or on a timer would mitigate that greatly.

(If you're wondering why that would be a common occurrence, I run QL via X11 forwarding. The connection will time out if I let the client sleep without quitting QL first, eventually causing QL to die on the host.)

If there's support for implementing this in core, I might have a go at it. Any implementation tips would be welcome, of course.

**Comment by bassstorm** [comment]
So, since this ticket is quite old, is that just about time effort or there's a serious technical limitation (e.g. massive refactoring) that doesn't allow to go with implementation?

**Comment by afontenot** [comment]
Unless I'm mistaken, this issue was fixed back in 2020 with a new plugin called "Auto Update Tags in Files". Maybe it should be the default behavior (if "Save ratings and play counts in tags" is checked), but anyone reading this who just wants this behavior, you can enable the plugin.

**Comment by declension** [comment]
Yes the plugin should be working. I think a lot of people agree longer term it should happen by default at which the point the plugin should not be necessary any more

**Comment by mmuffins** [comment]
I found this issue after losing a three-digit number of ratings for tracks in my library. What happened was that I'm currently importing a few music files from a different music player, and needed to ensure that the tags are consistent between both applications. At some point I highlighted all tracks by the same artist (which contained both the imported songs and older songs that were already in my library) in quod libet and clicked the `Rescan Songs` option from the context menu.

I did understand that this option would replace all the values in my library with data from the tags - but that's fine, because I enabled both the `Save ratings and play counts in tags` and `Auto-save tag changes` settings, so all changes I made in the past were already automatically persisted in the file tags. In addition I had the `Scan library on start` option enabled, so I would already have seen any discrepancies between files and the library after restarts in the past. 

I was aware of the `Auto Update Tags in Files` plugin but left it disabled because it seemed to first and foremost about keeping play counts in sync since all the options dealt with certain playback events, and I don't care about those. Even if I had it enabled, it might not have made much of a difference, because I tend to rate songs after I listened to them a few times, so I might not immediately listen to it again.

Maybe that's on me for not fully understanding the nuances of how quod libet handles persisting different file tags, but it simply never occurred to me the `Save ratings and play counts in tags` setting does not actually save ratings and play counts in tags and `Auto-save tag changes` does auto-save tags, but only if that tag is not the popularimeter tag. The current settings and behavior is misleading at best and broken at worst.

**Comment by bassstorm** [comment]
> I found this issue after losing a three-digit number of ratings for tracks in my library. What happened was that I'm currently importing a few music files from a different music player, and needed to ensure that the tags are consistent between both applications. At some point I highlighted all tracks by the same artist (which contained both the imported songs and older songs that were already in my library) in quod libet and clicked the `Rescan Songs` option from the context menu.
> 
> I did understand that this option would replace all the values in my library with data from the tags - but that's fine, because I enabled both the `Save ratings and play counts in tags` and `Auto-save tag changes` settings, so all changes I made in the past were already automatically persisted in the file tags. In addition I had the `Scan library on start` option enabled, so I would already have seen any discrepancies between files and the library after restarts in the past.
> 
> I was aware of the `Auto Update Tags in Files` plugin but left it disabled because it seemed to first and foremost about keeping play counts in sync since all the options dealt with certain playback events, and I don't care about those. Even if I had it enabled, it might not have made much of a difference, because I tend to rate songs after I listened to them a few times, so I might not immediately listen to it again.
> 
> Maybe that's on my for not fully understanding the nuances of how quod libet handles persisting different file tags, but it simply never occurred to me the `Save ratings and play counts in tags` setting does not actually save ratings and play counts in tags and `Auto-save tag changes` does auto-save tags, but only if that tag is not the popularimeter tag. The current settings and behavior is misleading at best and broken at worst.

I'd suggest you to try `fooyin` app. I use exactly same approach everything-is-file-metadata, so that effectively allows me to use `fooyin` and `foobar2000` interchangeably.

**Comment by mmuffins** [comment]
> I'd suggest you to try `fooyin` app. I use exactly same approach everything-is-file-metadata, so that effectively allows me to use `fooyin` and `foobar2000` interchangeably.

I do have fooyin installed. It's a great application, but still lacking a few features I am using which quod libet has. I also customized quod libet with a handfull of convenience scripts, which I cannot easily replace because the addon model is a bit more limited (and needs C which I have no clue about).
